### PR TITLE
fix(mcdu): Duplicate waypoints page callback

### PR DESCRIPTION
**Pull request by tracernz**
## Summary of Changes

The data manager sometimes returns the same waypoint/navaid more than once in result list. This was worked around previously by a filter on the duplicate names page, after we'd already decided to show that page which sometimes resulted in us showing that page with only one option (see https://github.com/flybywiresim/a32nx/pull/3752#issuecomment-817873934). I have implemented a filter in the FMC instead, so it can check properly the de-duplicated list before deciding if the duplicate names page is needed. I changed the criteria to the `icao` field matching as this identifies unique waypoints/navaids/airports/etc. in the MSFS database, rather than simply comparing locations.

Additionally there are quite a few pages that use GetOrSelect*ByIdent will not return to the correct page when the return key is pressed on the duplicate names page (I replaced the hardcoded f-pln with a function in #3752), so I have fixed those up (cc @St54Kevin, who pointed out the RADIO NAV page).

## Additional context

Discord username (if different from GitHub):

## Testing instructions

The duplicate names page is reachable by entering a waypoint or navaid in PROG, F-PLN, RAD NAV, DIR TO. Enter a navaid or waypoint that has duplicates at any of these places, and see the duplicate names page shows up, press the return key and make sure you get back to where you were. Enter DVR on the PROG page and make sure it doesn't show the duplicate names page, and does enter correctly (see example of the bug here https://youtu.be/SbbJH42vS-w?t=1964).

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 1]
```diff
@@ -1626,41 +1626,29 @@ class FMCMainDisplay extends BaseAirliners {
         return this._routeTripFuelWeight;
     }
 
-    getOrSelectVORsByIdent(ident, callback) {
-        this.dataManager.GetVORsByIdent(ident).then((navaids) => {
-            if (!navaids || navaids.length === 0) {
-                this.addNewMessage(NXSystemMessages.notInDatabase);
+    _getOrSelectWaypoints(getter, ident, callback) {
```
Comment (tracernz) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:1629]: re-factored common logic into one function here. Shifted responsibility for generating the not in database scratchpad message to the caller for VOR and NDB, as it already was for the catch-all waypoint function.

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_NavRadioPage.js [hunk 2]
```diff
@@ -143,14 +148,18 @@ class CDUNavRadioPage {
...
+                            mcdu.adf1IdIsPilotEntered = true;
+                            mcdu.adf1IdPilotValue = value;
+                            mcdu.adf1Frequency = navaids.infos.frequencyMHz;
+                            mcdu.radioNav.setADFActiveFrequency(1, mcdu.adf1Frequency);
+                            mcdu.requestCall(() => {
+                                CDUNavRadioPage.ShowPage(mcdu);
+                            });
+                        } else {
+                            mcdu.addNewMessage(NXSystemMessages.notInDatabase);
+                        }
```
Comment (tracernz) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_NavRadioPage.js:162]: Took the liberty of adding checks here. There may be some 3-letter idents not in the database.

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 3]
```diff
@@ -1626,41 +1626,29 @@ class FMCMainDisplay extends BaseAirliners {
...
+        getter(ident).then((waypoints) => {
+            console.log(waypoints);
```
Comment (davidwalschots) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:1631]: **question**: Is this logging intentional?

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 4]
```diff
@@ -1626,41 +1626,29 @@ class FMCMainDisplay extends BaseAirliners {
...
+            if (!waypoints || waypoints.length === 0) {
```
Comment (davidwalschots) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:1632]: **suggestion**: As `GetVORsByIdent`, `GetNDBsByIdent`, and `GetWaypointsByIdent` always return an `Array` the `length` check will suffice. Consider removing `!waypoints`
```suggestion
            if (waypoints.length === 0) {
```

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js (hunk 1 above)
Comment (davidwalschots) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:1629]: **thought**: I'm in doubt about having this function here. It feels like filtering is a concern of the `FMCDataManager`. Of course some other things this function does should be part of `FMCMainDisplay`.

I guess it's the somewhat "interesting" style of Asobo JavaScript coding that makes it hard to get this very clean right now.

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js (hunk 3 above)
Comment (tracernz) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:1631]: Haha, no, thanks. 😀

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js (hunk 1 above)
Comment (tracernz) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:1629]: I'll take a look and refactor it a little more. You're right that the data manager shouldn't send duplicates in the first place and it makes sense to filter then there for the benefit of other systems too.

**Comment by tracernz** [comment]
All comments addressed. The source of the duplicate waypoints turns out to be that some VORs (and maybe other radio navaids) are also intersections.

**Review by davidwalschots — APPROVED** [review]

**Comment by marcman86** [comment]
Category: QA Tester
Name: marcman86#4907
Date of testing: 25.04.2021
Version of the sim: 1.15.8.0
PR Tested: #4422
Tier of testing: 1
Changes to observe: Duplicate waypoints page

Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Turned on bat. and ext. power
3.) Tested duplicate waypoints page return
- PROG
- F-PLN
- RAD NAV
- DIR

4.) Entered DVR on PROG BRG / DIST
- Duplicate waypoints page does not show up

Issues: -

Overall Rating: Good

Conclusions of the testing: Works

Hope this helps :-)

**Comment by St54Kevin** [comment]
Category: QA Tester
Name: St54Kevin#3583
Date of testing: 25/04/2021
Version of the sim: 1.15.8.0
PR Tested: #4422 
Tier of testing: 1
Changes to observe: Duplicate waypoints page 

**Steps you've made to test the Pull Request:**
1. Loaded onto a Gate and powered up
2. Tested PROG, FPLN, RADNAV and DIRECT pages

**Media:**

**Issues (If Any):** - 

**Conclusions of the testing:** LGTM. Works as expected. Thanks for adressing this

**Review by davidwalschots — APPROVED** [review]

**Pull request merged**